File: test/pgo/parser/ModularPlusCalParserTest.java
package pgo.parser;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static pgo.model.mpcal.ModularPlusCalBuilder.mpcal;
import static pgo.model.pcal.PlusCalBuilder.*;
import static pgo.model.tla.TLABuilder.*;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import pgo.model.mpcal.ModularPlusCalBlock;
import pgo.model.pcal.PlusCalFairness;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@RunWith(Parameterized.class)
public class ModularPlusCalParserTest {
	@Parameterized.Parameters
	public static List<Object[]> data() {
		return Arrays.asList(new Object[][] {
				{"---- MODULE Test ----\n" +
						"(* --mpcal Test {\n" +
						"     macro M(a) {\n" +
						"       print a;\n" +
						"     }\n" +
						"     procedure P(b) {\n" +
						"       print b;\n" +
						"     }\n" +
						"     variables global1 = 1, global2 = 2;\n" +
						"     {\n" +
						"       l1: print <<global1, global2>>;\n" +
						"     }\n" +
						"}\n" +
						"*)",
						mpcal("Test",
								Collections.emptyList(), Collections.singletonList(macro("M", Collections.singletonList("a"), printS(idexp("a")))), Collections.singletonList(procedure(
										"P",
										Collections.singletonList(pcalVarDecl("b", false, false, PLUSCAL_DEFAULT_INIT_VALUE)),
										Collections.emptyList(),
										printS(idexp("b")))), Collections.emptyList(), Collections.emptyList(), Arrays.asList(
										pcalVarDecl("global1", false, false, num(1)),
										pcalVarDecl("global2", false, false, num(2))),
								Collections.emptyList(),
								Collections.singletonList(labeled(
										label("l1"),
										printS(tuple(idexp("global1"), idexp("global2")))
								))
						)
				},
				{"---- MODULE Test ----\n" +
						"(* --mpcal Test {\n" +
						"     macro M(a) {\n" +
						"       print a;\n" +
						"     }\n" +
						"     procedure P(b) {\n" +
						"       print b;\n" +
						"     }\n" +
						"     variables global1 = 1, global2 = 2;\n" +
						"     process (P \\in 1..3) {\n" +
						"       l1: print <<global1, global2>>;\n" +
						"     }\n" +
						"}\n" +
						"*)",
						mpcal("Test",
								Collections.emptyList(),
								Collections.singletonList(macro(
										"M",
										Collections.singletonList("a"),
										printS(idexp("a")))),
								Collections.singletonList(procedure(
										"P",
										Collections.singletonList(pcalVarDecl("b", false, false, PLUSCAL_DEFAULT_INIT_VALUE)),
										Collections.emptyList(),
										printS(idexp("b")))), Collections.emptyList(), Collections.emptyList(), Arrays.asList(
										pcalVarDecl("global1", false, false, num(1)),
										pcalVarDecl("global2", false, false, num(2))),
								Collections.emptyList(),
								process(
										pcalVarDecl("P", false, true, binop("..", num(1), num(3))),
										PlusCalFairness.UNFAIR,
										Collections.emptyList(),
										labeled(
												label("l1"),
												printS(tuple(idexp("global1"), idexp("global2")))
										)
								)
						)
				},
		});
	}

	private String blockString;
	private ModularPlusCalBlock blockExpected;

	public ModularPlusCalParserTest(String blockString, ModularPlusCalBlock blockExpected) {
		this.blockString = blockString;
		this.blockExpected = blockExpected;
	}

	private static final Path testFile = Paths.get("TEST");

	@Test
	public void test() throws ParseFailureException {
		LexicalContext ctx = new LexicalContext(
				testFile,
				String.join(System.lineSeparator(), blockString.split("\n")));
		System.out.println(blockString);
		ModularPlusCalBlock block = ModularPlusCalParser.readBlock(ctx);
		assertThat(block, is(blockExpected));
	}
}

File: test/pgo/parser/ModularPlusCalParserTest.scala
package pgo.parser

import org.scalatest.funsuite.AnyFunSuite
import pgo.TestingUtils
import pgo.model.mpcal.ModularPlusCalBlock
import pgo.model.mpcal.ModularPlusCalBuilder._
import pgo.model.pcal.PlusCalBuilder._
import pgo.model.pcal.{PlusCalFairness, PlusCalStatement}
import pgo.model.tla.TLABuilder._

import scala.jdk.CollectionConverters._

class ModularPlusCalParserTest extends AnyFunSuite {
  def check(tag: String)(pair: (String,ModularPlusCalBlock)): Unit =
    test(tag) {
      val(specStr,expectedBlock) = pair
      val (_, _, actualBlock) = TestingUtils.parseMPCalFromString(specStr)
      assert(actualBlock == expectedBlock)
    }

  check("single-process with procedure and macro") {
    """---- MODULE Test ----
      |(* --mpcal Test {
      |     macro M(a) {
      |       print a;
      |     }
      |     procedure P(b) {
      |       print b;
      |     }
      |     variables global1 = 1, global2 = 2;
      |     {
      |       l1: print <<global1, global2>>;
      |     }
      |}
      |*)
      |\* BEGIN TRANSLATION
      |====""".stripMargin -> mpcal(
      "Test",
      Nil.asJava,
      List(
        `macro`("M",
          List("a").asJava,
          printS(idexp("a")))).asJava,
      List(
        procedure("P",
          List(pcalVarDecl("b", false, false, PLUSCAL_DEFAULT_INIT_VALUE)).asJava,
          Nil.asJava,
          printS(idexp("b")))).asJava,
      Nil.asJava,
      Nil.asJava,
      List(
        pcalVarDecl("global1", false, false, num(1)),
        pcalVarDecl("global2", false, false, num(2))).asJava,
      Nil.asJava,
      List[PlusCalStatement](
        labeled(label("l1"),
          printS(tuple(idexp("global1"), idexp("global2"))))).asJava)
  }

  check("multi-process with procedure and macro") {
    """---- MODULE Test ----
      |EXTENDS Integers
      |(* --mpcal Test {
      |     macro M(a) {
      |       print a;
      |     }
      |     procedure P(b) {
      |       print b;
      |     }
      |     variables global1 = 1, global2 = 2;
      |     process (P \in 1..3) {
      |       l1: print <<global1, global2>>;
      |     }
      |}
      |*)
      |\* BEGIN TRANSLATION
      |====
      |""".stripMargin -> mpcal(
      "Test",
      Nil.asJava,
      List(
        `macro`("M",
          List("a").asJava,
          printS(idexp("a")))).asJava,
      List(
        procedure("P",
          List(pcalVarDecl("b", false, false, PLUSCAL_DEFAULT_INIT_VALUE)).asJava,
          Nil.asJava,
          printS(idexp("b")))).asJava,
      Nil.asJava,
      Nil.asJava,
      List(
        pcalVarDecl("global1", false, false, num(1)),
        pcalVarDecl("global2", false, false, num(2))).asJava,
      Nil.asJava,
      process(
        pcalVarDecl("P", false, true, binop("..", num(1), num(3))),
        PlusCalFairness.UNFAIR,
        Nil.asJava,
        labeled(label("l1"),
          printS(tuple(idexp("global1"), idexp("global2"))))))
  }
}
